### PR TITLE
Use coding style.

### LINKED ISSUE #3654: Coding standars link on the 'contributing to Solidity' page errors to a 404
# Issue 
When clicking on the link 'coding standards' in the "workflow for pull requests" section on the *Contributing* to Solidity page found [here](https://solidity.readthedocs.io/en/latest/contributing.html) it errors on a 404.  

## Steps to recreate 

1.  Navigate to the Contributing page found [here](https://solidity.readthedocs.io/en/latest/contributing.html)
2.  Click on "coding standards" which links to the following [url](https://raw.githubusercontent.com/ethereum/cpp-ethereum/develop/CodingStandards.txt)
3.  Receive 404 

## Additional details 
Broken link leads to the following url which results in a 404 

`https://solidity.readthedocs.io/en/latest/contributing.html`

## Specs 

1. Using High Sierra 10.13.2 
2. Using Safari 11.0.2  

Ideally, it should link to our coding standards documentation.  

## Ideal solution 
Page links to our coding  standards

### PR DESCRIPTION
**Pull request by chriseth**
Fixes #3654.

### CODE REVIEW

File: CODING_STYLE.md [hunk 1]
```diff
@@ -0,0 +1,254 @@
+0. Formatting
+
+GOLDEN RULE: Follow the style of the existing code when you make changes.
+
+a. Use tabs for leading indentation
+- tab stops are every 4 characters.
```
Comment (chriseth) [inline CODING_STYLE.md]: Proposed change: add: `(only relevant for line length)`

File: CODING_STYLE.md [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+- One indentation level -> exactly one byte (i.e. a tab character) in the source file.
+- If you have run-on lines, indent as you would for a block.
```
Comment (chriseth) [inline CODING_STYLE.md]: Proposed change: Remove this line.

File: CODING_STYLE.md [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+b. Line widths:
+- Don't worry about having lines of code > 80-char wide.
```
Comment (chriseth) [inline CODING_STYLE.md]: Proposed change: Lines should be at most 100 characters wide to make help with diffs and reduce merge conflicts.

File: CODING_STYLE.md [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+- Lines of comments should be formatted according to ease of viewing, but simplicity is to be preferred over beauty.
+c. Don't use braces for condition-body one-liners.
```
Comment (chriseth) [inline CODING_STYLE.md]: Proposed change: `Single-statement blocks should not have braces, unless required for clarity.`

File: CODING_STYLE.md [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+d. Never place condition bodies on same line as condition.
+e. Space between first paren and keyword, but *not* following first paren or preceding final paren.
```
Comment (chriseth) [inline CODING_STYLE.md]: Proposed change: `Space between keyword and opening parenthesis, but not following opening parenthesis or before final parenthesis.`

File: CODING_STYLE.md [hunk 6]
```diff
@@ -0,0 +1,254 @@
...
+f. No spaces when fewer than intra-expression three parens together; when three or more, space according to clarity.
```
Comment (chriseth) [inline CODING_STYLE.md]: Propose to remove.

File: CODING_STYLE.md [hunk 7]
```diff
@@ -0,0 +1,254 @@
...
+g. No spaces for subscripting or unary operators.
```
Comment (chriseth) [inline CODING_STYLE.md]: Proposed change: `substricting -> index`

File: CODING_STYLE.md [hunk 8]
```diff
@@ -0,0 +1,254 @@
...
+h. No space before ':' but one after it, except in the ternary operator: one on both sides.
+i. Space all other operators.
```
Comment (chriseth) [inline CODING_STYLE.md]: add `around` and `binary`

Perhaps also add: no spaces around  `->` or `.`

File: CODING_STYLE.md [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+j. Braces, when used, always have their own lines and are at same indentation level as "parent" scope.
```
Comment (chriseth) [inline CODING_STYLE.md]: Add: If lines are broken, a list of elements enclosed with parentheses (of any kind) and separated by a separator (of any kind) are formatted such that there is exactly one element per line, followed by the separator, the opening parenthesis is on the first line, followed by a line break and the closing parenthesis is on a line of its own (unindented).

Example:

File: CODING_STYLE.md [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+    anotherLongVariableName
+);
+
+
+
+1. Namespaces;
+
+a. No "using namespace" declarations in header files.
+b. All symbols should be declared in a namespace except for final applications.
+c. Preprocessor symbols should be prefixed with the namespace in all-caps and an underscore.
```
Comment (chriseth) [inline CODING_STYLE.md]: Add: Use anonymous namespaces for helpers whose scope is a cpp file only.

File: CODING_STYLE.md [hunk 11]
```diff
@@ -0,0 +1,254 @@
...
+All other entities' first alpha is lower case.
+
+
+
+4. Variable prefixes:
+
+a. Leading underscore "_" to parameter names.
+- Exception: "o_parameterName" when it is used exclusively for output. See 6(f).
+- Exception: "io_parameterName" when it is used for both input and output. See 6(f).
+b. Leading "c_" to const variables (unless part of an external API).
```
Comment (chriseth) [inline CODING_STYLE.md]: Propose to remove.

File: CODING_STYLE.md [hunk 12]
```diff
@@ -0,0 +1,254 @@
...
+c. Leading "g_" to global (non-const) variables.
+d. Leading "s_" to static (non-const, non-global) variables.
+
+
+
+5. Error reporting:
+
+- Prefer exception to bool/int return type.
```
Comment (chriseth) [inline CODING_STYLE.md]: Propose to remove.

Comment (chriseth) [reply CODING_STYLE.md]: (error reporting is done in a special way in solidity)

We could add a rule to use `solAssert()` generously to check assumptions that cross multiple files. Use `solAssert()` each time you dereference a pointer.

File: CODING_STYLE.md [hunk 13]
```diff
@@ -0,0 +1,254 @@
...
+
+
+
+6. Declarations:
+
+a. {Typename} + {qualifiers} + {name}.
+b. Only one per line.
+c. Associate */& with type, not variable (at ends with parser, but more readable, and safe if in conjunction with (b)).
+d. Favour declarations close to use; don't habitually declare at top of scope ala C.
+e. Always pass non-trivial parameters with a const& suffix.
```
Comment (chriseth) [inline CODING_STYLE.md]: Propose to add: Pass non-trivial parameters as const reference, unless the data is to be copied into the function, then either pass by const ref or by value and use std::move.

File: CODING_STYLE.md [hunk 14]
```diff
@@ -0,0 +1,254 @@
...
+f. If a function returns multiple values, use std::tuple (std::pair acceptable). Prefer not using */& arguments, except where efficiency requires.
```
Comment (chriseth) [inline CODING_STYLE.md]: or better: introduce a struct type.

Comment (chriseth) [reply CODING_STYLE.md]: Add: Use pointer arguments if the value can be a nullptr, otherwise always use reference.

File: CODING_STYLE.md [hunk 15]
```diff
@@ -0,0 +1,254 @@
...
+g. Never use a macro where adequate non-preprocessor C++ can be written.
+h. Make use of auto whenever type is clear or unimportant:
+- Always avoid doubly-stating the type.
+- Use to avoid vast and unimportant type declarations.
+- However, avoid using auto where type is not immediately obvious from the context, and especially not for arithmetic expressions.
```
Comment (chriseth) [inline CODING_STYLE.md]: I would say this is almost always the case, so propose to remove `h`.

File: CODING_STYLE.md [hunk 16]
```diff
@@ -0,0 +1,254 @@
...
+f. For a property 'foo'
+- Member: m_foo;
+- Getter: foo() [ also: for booleans, isFoo() ];
+- Setter: setFoo();
+
+
+
+9. Naming
+
+a. Collection conventions:
```
Comment (chriseth) [inline CODING_STYLE.md]: Propose toremove.

File: CODING_STYLE.md [hunk 17]
```diff
@@ -0,0 +1,254 @@
...
+- -s means std::vector e.g. using MyTypes = std::vector<MyType>
+- -Set means std::set e.g. using MyTypeSet = std::set<MyType>
+- -Hash means std::unordered_set e.g. using MyTypeHash = std::unordered_set<MyType>
+b. Class conventions:
```
Comment (chriseth) [inline CODING_STYLE.md]: Propose to remove.

File: CODING_STYLE.md [hunk 18]
```diff
@@ -0,0 +1,254 @@
...
+- -Face means the interface of some shared concept. (e.g. FooFace might be a pure virtual class.)
+c. Avoid unpronouncable names;
+- If you need to shorten a name favour a pronouncable slice of the original to a scattered set of consonants.
+- e.g. Manager shortens to Man rather than Mgr.
```
Comment (chriseth) [inline CODING_STYLE.md]: Propose to remove, perhaps also add: names should be shortened only if they are extremely common, but shortening should be generally avoided.

File: CODING_STYLE.md [hunk 19]
```diff
@@ -0,0 +1,254 @@
...
+
+10. Type-definitions
+
+a. Prefer 'using' to 'typedef'. e.g. using ints = std::vector<int>; rather than typedef std::vector<int> ints;
+b. Generally avoid shortening a standard form that already includes all important information:
+- e.g. stick to shared_ptr<X> rather than shortening to ptr<X>.
+c. Where there are exceptions to this (due to excessive use and clear meaning), note the change prominently and use it consistently.
+- e.g. using Guard = std::lock_guard<std::mutex>; ///< Guard is used throughout the codebase since it's clear in meaning and used commonly. 
+d. In general expressions should be roughly as important/semantically meaningful as the space they occupy.
+
```
Comment (chriseth) [inline CODING_STYLE.md]: Add: Avoid naming types unless they are rather complicated. Consider the number of items a brain can keep track of at the same time.

File: CODING_STYLE.md [hunk 20]
```diff
@@ -0,0 +1,254 @@
...
+11. Commenting
+
+a. Comments should be doxygen-compilable, using @notation rather than \notation.
+b. Document the interface, not the implementation.
+- Documentation should be able to remain completely unchanged, even if the method is reimplemented.
+- Comment in terms of the method properties and intended alteration to class state (or what aspects of the state it reports).
+- Be careful to scrutinise documentation that extends only to intended purpose and usage.
+- Reject documentation that is simply an English transaction of the implementation.
+
```
Comment (chriseth) [inline CODING_STYLE.md]: Avoid in-code comments. Instead, try to extract blocks of functionality into functions. This often already eliminates the need for an in-code comment.

File: CODING_STYLE.md [hunk 21]
```diff
@@ -0,0 +1,254 @@
...
+
+
+12. Include Headers
+
+Includes should go in increasing order of generality (libethereum -> libethcore -> libdevcrypto -> libdevcore -> boost -> STL).  For example:
```
Comment (chriseth) [inline CODING_STYLE.md]: Add: The corresponding .h file should be the first include in the respective .cpp file.

File: CODING_STYLE.md [hunk 22]
```diff
@@ -0,0 +1,254 @@
...
+
+#include <libethereum/Defaults.h>
+#include <libdevcrypto/SHA3.h>
+#include <libdevcore/Log.h>
+#include <libdevcore/Exceptions.h>
+#include <libdevcore/CommonData.h>
+#include <libdevcore/Common.h>
+#include <boost/filesystem.hpp>
+#include <string>
```
Comment (chriseth) [inline CODING_STYLE.md]: Separate groups of includes by empty lines.

File: CODING_STYLE.md [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+
+See http://stackoverflow.com/questions/614302/c-header-order/614333#614333 for the reason: this makes it easier to find missing includes in header files.
+
+
+
+13. Logging
```
Comment (chriseth) [inline CODING_STYLE.md]: Remove, we do not log.

File: CODING_STYLE.md (hunk 3 above)
Comment (axic) [reply CODING_STYLE.md]: Perhaps use the same rule as for .sol, 96 or something like that, was merged last week.

File: CODING_STYLE.md (hunk 21 above)
Comment (axic) [reply CODING_STYLE.md]: Also the example list should be updated for solidity.

File: CODING_STYLE.md (hunk 15 above)
Comment (axic) [reply CODING_STYLE.md]: What do you mean by "this is always the case"?

File: CODING_STYLE.md (hunk 14 above)
Comment (axic) [reply CODING_STYLE.md]: Perhaps you want to mention when `boost::optional` is appropriate.

File: CODING_STYLE.md (hunk 12 above)
Comment (axic) [reply CODING_STYLE.md]: Also mention `solUnimplementedAssert`.

File: CODING_STYLE.md (hunk 11 above)
Comment (axic) [reply CODING_STYLE.md]: You mean `b.` or both `a.` and `b.`?

File: CODING_STYLE.md (hunk 9 above)
Comment (axic) [reply CODING_STYLE.md]: Example is missing.

File: CODING_STYLE.md (hunk 15 above)
Comment (chriseth) [reply CODING_STYLE.md]: The type is almost always not immediately obvious from the context. Perhaps reframe this as 
```
Avoid using auto when the type is not mentioned in the direct vincinity.
Example of good use:
auto const& x = dynamic_cast<SomeType const&>(bla);
but should actually be
SomeType const& x = dynamic_cast<decltype(x)>(bla);
to help IDEs also find the type.
```

File: CODING_STYLE.md (hunk 9 above)
Comment (chriseth) [reply CODING_STYLE.md]: The example is further down.

Comment (chriseth) [reply CODING_STYLE.md]: will add `(see example below)`

File: CODING_STYLE.md (hunk 11 above)
Comment (chriseth) [reply CODING_STYLE.md]: only b, but another rule actually says that cases like `o_` and `io_` should not be used.

File: CODING_STYLE.md (hunk 14 above)
Comment (chriseth) [reply CODING_STYLE.md]: Good idea!

**Comment by chriseth** [comment]
@axic @bit-shift @ekpyron we should get this merged, we can always make changes later.

**Comment by chriseth** [comment]
Merging for now.

**Pull request merged**